core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannel.scala

```diff
@@ -0,0 +1,93 @@
+  private val pendingTxnMap: concurrent.Map[Long, TransactionMetadata] = concurrent.TrieMap.empty[Long, TransactionMetadata]
+
+  def addNewBroker(broker: Node) {
+    val markersQueue = new LinkedBlockingQueue[CoordinatorEpochAndMarkers]()
+    brokerStateMap.put(broker.id, DestinationBrokerAndQueuedMarkers(broker, markersQueue))
+    trace(s"Added destination broker ${broker.id}: $broker")
+  }
+
+  private[transaction] def addRequestForBroker(brokerId: Int, txnMarkerRequest: CoordinatorEpochAndMarkers) {
+    val markersQueue = brokerStateMap(brokerId).markersQueue
+    markersQueue.add(txnMarkerRequest)
+    trace(s"Added markers $txnMarkerRequest for broker $brokerId")
+  }
+
+  def addRequestToSend(pid: Long, epoch: Short, result: TransactionResult, coordinatorEpoch: Int, topicPartitions: immutable.Set[TopicPartition]): Unit = {
+    val partitionsByDestination: immutable.Map[Int, immutable.Set[TopicPartition]] = topicPartitions.groupBy { topicPartition: TopicPartition =>
+      val leaderForPartition = metadataCache.getPartitionInfo(topicPartition.topic, topicPartition.partition)
+      leaderForPartition match {
+        case Some(partitionInfo) =>
+          val brokerId = partitionInfo.leaderIsrAndControllerEpoch.leaderAndIsr.leader
+          if (!brokerStateMap.contains(brokerId)) {
+            val broker = metadataCache.getAliveEndpoint(brokerId, interBrokerListenerName).get
+            addNewBroker(broker)
+          }
+          brokerId
+        case None =>
+          // TODO: there is a rare case that the producer gets the partition info from another broker who has the newer information of the
+          // partition, while the TC itself has not received the propagated metadata update; do we need to block when this happens?
+          throw new IllegalStateException("Cannot find the leader broker for the txn marker to write for topic partition " + topicPartition)
+      }
+    }
+
+    for ((brokerId: Int, topicPartitions: immutable.Set[TopicPartition]) <- partitionsByDestination) {
+      val txnMarker = new TxnMarkerEntry(pid, epoch, result, topicPartitions.toList.asJava)
+      addRequestForBroker(brokerId, CoordinatorEpochAndMarkers(coordinatorEpoch, Utils.mkList(txnMarker)))
+    }
+  }
+
+  def maybeAddPendingRequest(metadata: TransactionMetadata): Unit = {
+    val existingMetadataToWrite = pendingTxnMap.putIfAbsent(metadata.pid, metadata)
+    if (existingMetadataToWrite.isDefined)
+      throw new IllegalStateException(s"There is already a pending txn to write its markers ${existingMetadataToWrite.get}; this should not happen")
+  }
+
+  def removeCompletedTxn(pid: Long): Unit = {
+    pendingTxnMap.remove(pid)
+  }
+
+  def pendingTxnMetadata(pid: Long): TransactionMetadata = {
+    pendingTxnMap.getOrElse (pid, throw new IllegalStateException ("Cannot find the pending txn marker in cache; it should not happen") )
+  }
+
+  def clear(): Unit = {
```

Comment:
Yes, only clearing transactionalIds where the partition has emigrated. For in-flight-requests, we probably want to return a NotTransactionCoordinator error.